### PR TITLE
[dist] fix: correct FSDP2 grad norm reduction

### PR DESCRIPTION
**Pull request by Zx55**
## Summary

Fix two FSDP2 gradient norm issues:

- Under HSDP, reduce shard-local norm statistics over the `dp_shard` group instead of the full FSDP group. Each shard group is replicated across `dp_replicate`, so reducing over the full group over-counts the norm. For replicated but unsharded FSDP2, skip the extra norm reduction.
- Compute the local p-norm sum in bounded chunks using `torch._foreach_norm`. This avoids materializing all gradients as fp32 at once while keeping kernel-launch overhead bounded.

## Validation

```bash
.venv/bin/python -m pytest -q tests/utils/test_fsdp2_clip_grad_norm.py
```

Result:

```text
5 passed
```

```bash
.venv/bin/python -m py_compile \
  veomni/distributed/fsdp2/clip_grad_norm.py \
  tests/utils/test_fsdp2_clip_grad_norm.py
```

### CODE REVIEW

File: veomni/distributed/fsdp2/clip_grad_norm.py [hunk 1]
```diff
@@ -172,24 +179,18 @@ def _raise_if_nonfinite(total_norm: torch.Tensor, norm_type: float, error_if_non
...
+        for param in params:
+            g = param.grad
+            if g is None:
+                continue
+            if isinstance(g, DTensor):
+                g = g.to_local()
+            g = g.detach().to(torch.float32)
+            gn = torch.norm(g, p=p)
+            res = res + (gn**p).to(reduce_device)
     return res
```
[the remark this answers is not in the document]
Comment (Zx55) [reply veomni/distributed/fsdp2/clip_grad_norm.py:222]: Addressed in f538440f: `_local_pth_sum` now uses bounded chunks with `torch._foreach_norm`, grouped by device, so it avoids all-grad fp32 materialization while reducing per-parameter kernel launch overhead.

File: tests/utils/test_fsdp2_clip_grad_norm.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+def test_fsdp_grad_norm_reduce_groups_use_fsdp_group_for_plain_fsdp2():
+    ps = _parallel_state()
+
+    assert _fsdp_grad_norm_reduce_groups(ps) == [("fsdp", ps.fsdp_group)]
+
+
+def test_fsdp_grad_norm_reduce_groups_use_shard_group_for_hsdp():
+    ps = _parallel_state(dp_replicate_enabled=True, dp_shard_size=4)
+
+    assert _fsdp_grad_norm_reduce_groups(ps) == [("fsdp_shard", ps.dp_shard_group)]
```
[the remark this answers is not in the document]
Comment (Zx55) [reply tests/utils/test_fsdp2_clip_grad_norm.py:39]: Addressed in f538440f: added `test_fsdp_grad_norm_reduce_groups_skip_replicated_unsharded_fsdp2` for the replicated-but-unsharded case.

File: veomni/distributed/fsdp2/clip_grad_norm.py [hunk 3]
```diff
@@ -51,6 +46,19 @@ def clip_grad_norm(
...
+    """Process groups for FSDP2 grad-norm reduction.
+
+    Under HSDP each shard group is replicated across ``dp_replicate``; only sum
+    shard-local p-th powers across ``dp_shard``, not across replicas.
+    """
+    if ps.dp_mode != "fsdp2":
+        return []
+    if ps.dp_replicate_enabled and ps.dp_shard_size > 1:
+        return [("fsdp_shard", ps.dp_shard_group)]
+    return [("fsdp", ps.fsdp_group)]
```
[the remark this answers is not in the document]
Comment (Zx55) [reply veomni/distributed/fsdp2/clip_grad_norm.py]: Addressed in f538440f: when `dp_replicate_enabled` is true and `dp_shard_size == 1`, `_fsdp_grad_norm_reduce_groups` now returns `[]`, and the unit test covers this case.

**Comment by Luosuu** [comment]
@Zx55 Thanks for the contribution. I will take a look soon

File: veomni/distributed/fsdp2/clip_grad_norm.py [hunk 4]
```diff
@@ -51,6 +47,19 @@ def clip_grad_norm(
...
+def _fsdp_grad_norm_reduce_groups(ps) -> List[tuple[str, dist.ProcessGroup | None]]:
+    """Process groups for FSDP2 grad-norm reduction.
+
+    Under HSDP each shard group is replicated across ``dp_replicate``; only sum
+    shard-local p-th powers across ``dp_shard``, not across replicas.
+    """
+    if ps.dp_mode != "fsdp2":
+        return []
+    if ps.dp_replicate_enabled:
+        return [("fsdp_shard", ps.dp_shard_group)] if ps.dp_shard_size > 1 else []
```
Comment (Luosuu) [inline veomni/distributed/fsdp2/clip_grad_norm.py]: I think we should include sp here as well. When HSDP+SP is enabled, the weight sharding dim is `dp_shard_sp` instead of `dp_shard`. For example, when fsdp size=8, sp=2, dp_shard_sp_size is 8 and dp_shard_size is 4.

File: veomni/distributed/fsdp2/clip_grad_norm.py [hunk 5]
```diff
@@ -172,24 +180,36 @@ def _raise_if_nonfinite(total_norm: torch.Tensor, norm_type: float, error_if_non
...
+
+    chunks: dict[torch.device, list[torch.Tensor]] = {}
+
+    def flush_chunk(device: torch.device) -> None:
+        nonlocal res
+        chunk = chunks.get(device)
+        if not chunk:
+            return
+        chunk_fp32 = [g.to(torch.float32) for g in chunk]
+        norms = torch._foreach_norm(chunk_fp32, p)
```
Comment (Luosuu) [inline veomni/distributed/fsdp2/clip_grad_norm.py]: I think we need to preserve the _has_foreach_support check for compatibility

**Review by Luosuu — CHANGES_REQUESTED** [review]
Thanks again, I have some comments

**Comment by Zx55** [comment]
Hi @Luosuu, thanks for the comments. I updated the reduction group to use `dp_shard_sp` when SP participates in FSDP sharding, and restored the `_has_foreach_support` fallback in `_local_pth_sum()` while keeping the bounded chunking to avoid fp32 grad materialization OOMs.

**Review by Luosuu — APPROVED** [review]
